src/server/builder.ts

```diff
@@ -319,18 +324,19 @@ namespace ts.server {
         getFilesAffectedBy(scriptInfo: ScriptInfo): string[] {
             this.ensureProjectDependencyGraphUpToDate();
 
+            const singleFileResult = scriptInfo.hasMixedContent ? [] : [scriptInfo.fileName];
             const fileInfo = this.getFileInfo(scriptInfo.path);
             if (!fileInfo || !fileInfo.updateShapeSignature()) {
-                return [scriptInfo.fileName];
+                return singleFileResult;
             }
 
             if (!fileInfo.isExternalModuleOrHasOnlyAmbientExternalModules()) {
-                return this.project.getFileNamesWithoutDefaultLib();
+                return map(filter(this.project.getScriptInfosWithoutDefaultLib(), shouldEmitFile), info => info.fileName);

```

Comment:
Actuall y`getScriptInfosWithoutDefaultLib` is also implemented using `filter` but I don't think it is used anywhere else except for these two places. Can you have method `Project.getAllEmittableFiles()` that will internally fuse `filter(is not default lib)` + `filter (should emit)` + `map (get file name)` into one loop.